### ISSUE TITLE
No internet after authentication

### ISSUE THREAD

**Issue by NourEddineX** [openNDS/openNDS#94]
First: I want to thank the team working on this project, and all the contributors.
Second:
### Issue description
Evern after being authenticated, I'm not able to access the internet, I suspected that my isp is trying to block traffic because of fwmark, but this doesn't look true since I'm not even able to reach the nexthop after opennds (which is my home dsl router).

I'm using OpenNDS on debian buster x86 (kernel 4.19), with hostapd, dnsmasq for dhcp and dns, and networking.service for IP network management, I developed my own custom captive portal based on /usr/lib/opennds/login.sh to use voucher code instead of username and password, the code is still dirty but it works

### Debuging info
opennds.conf
```
GatewayName Gateway
GatewayInterface wlan0
gatewayfqdn gateway.local
login_option_enabled 2
use_outdated_mhd 0
unescape_callback_enabled 0
FirewallRuleSet authenticated-users {
  FirewallRule allow all
}
FirewallRuleSet preauthenticated-users {
FirewallRule allow udp port 67
FirewallRule allow udp port 53
FirewallRule allow tcp port 53	
}
FirewallRuleSet users-to-router {
   FirewallRule allow udp port 53	
   FirewallRule allow tcp port 53	
   FirewallRule allow udp port 67
   FirewallRule allow tcp port 22
   FirewallRule allow tcp port 80
   FirewallRule allow tcp port 443
}
MaxClients 1000
TrustedMACList <REDACTED MAC 3>, <REDACTED MAC 2>
ratecheckwindow 2
```
```
$ ndsctl status
==================
openNDS Status
====
Version: 8.0.0
Uptime: 54s
Gateway Name: Gateway
Managed interface: wlan0
Managed IP range: 0.0.0.0/0
Server listening: http://<REDACTED openNDS IP>:2050
Binauth: Disabled
Preauth Script: /usr/lib/opennds/login.sh
FAS: Secure Level 1, URL: http://<REDACTED openNDS IP>:2050/opennds_preauth/
Client Check Interval: 30s
Rate Check Window: 2 check intervals (60s)
Preauthenticated Client Idle Timeout: 10m
Authenticated Client Idle Timeout: 120m
Download rate limit (default per client): no limit
Upload rate limit (default per client): no limit
Download quota (default per client): no limit
Upload quota (default per client): no limit
Total download: 98 kByte; avg: 14.59 kbit/s
Total upload: 42 kByte; avg: 6.31 kbit/s
====
Client authentications since start: 1
Current clients: 1

Client 0
  IP: <REDACTED IP 1> MAC: <REDACTED MAC 1>
  Last Activity: Sat Jan 09 05:46:56 2021 (0s ago)
  Session Start: Sat Jan 09 05:46:45 2021 (11s ago)
  Session End:   Sat Jan 09 05:56:45 2021 (9m 49s left)
  Token: ae642c16
  State: Authenticated
  Upload rate limit: 1000 kb/s
  Download rate limit: 2000 kb/s
  Upload quota: 500000 kB
  Download quota: 500000 kB
  Upload this session: 5 kB; Session avg: 4.10 kb/s
  Download this session: 13 kB; Session avg: 10.06 kb/s

====
Blocked MAC addresses: none
Allowed MAC addresses: N/A
Trusted MAC addresses:
  <REDACTED MAC 2>
  <REDACTED MAC 3>
Walled Garden FQDNs: none
Walled Garden Ports: none
========
```

```
$ iptables-save
# Generated by iptables-save v1.8.2 on Sat Jan  9 05:51:46 2021
*filter
:INPUT ACCEPT [154:15212]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [437:133263]
:ndsAUT - [0:0]
:ndsNET - [0:0]
:ndsRTR - [0:0]
:ndsTRT - [0:0]
:ndsTRU - [0:0]
-A INPUT -i wlan0 -j ndsRTR
-A FORWARD -i wlan0 -j ndsNET
-A FORWARD -i eth0 -o wlan0 -m state --state RELATED,ESTABLISHED -j ACCEPT
-A FORWARD -i wlan0 -o eth0 -j ACCEPT
-A ndsAUT -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A ndsAUT -j ACCEPT
-A ndsAUT -j REJECT --reject-with icmp-port-unreachable
-A ndsNET -m mark --mark 0x10000/0x30000 -j DROP
-A ndsNET -m conntrack --ctstate INVALID -j DROP
-A ndsNET -p tcp -m tcp --tcp-flags SYN,RST SYN -j TCPMSS --clamp-mss-to-pmtu
-A ndsNET -m mark --mark 0x20000/0x30000 -j ACCEPT
-A ndsNET -m mark --mark 0x30000/0x30000 -j ndsAUT
-A ndsNET -p udp -m udp --dport 67 -j ACCEPT
-A ndsNET -p udp -m udp --dport 53 -j ACCEPT
-A ndsNET -p tcp -m tcp --dport 53 -j ACCEPT
-A ndsNET -j REJECT --reject-with icmp-port-unreachable
-A ndsRTR -m mark --mark 0x10000/0x30000 -j DROP
-A ndsRTR -m conntrack --ctstate INVALID -j DROP
-A ndsRTR -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A ndsRTR -p tcp -m tcp ! --tcp-option 2 --tcp-flags SYN SYN -j DROP
-A ndsRTR -p tcp -m tcp --dport 2050 -j ACCEPT
-A ndsRTR -p tcp -m tcp --dport 2050 -j ACCEPT
-A ndsRTR -m mark --mark 0x20000/0x30000 -j ACCEPT
-A ndsRTR -p udp -m udp --dport 53 -j ACCEPT
-A ndsRTR -p tcp -m tcp --dport 53 -j ACCEPT
-A ndsRTR -p udp -m udp --dport 67 -j ACCEPT
-A ndsRTR -p tcp -m tcp --dport 22 -j ACCEPT
-A ndsRTR -p tcp -m tcp --dport 80 -j ACCEPT
-A ndsRTR -p tcp -m tcp --dport 443 -j ACCEPT
-A ndsRTR -j REJECT --reject-with icmp-port-unreachable
COMMIT
# Generated by iptables-save v1.8.2 
*mangle
:PREROUTING ACCEPT [405:33179]
:INPUT ACCEPT [142:15471]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [186:42883]
:POSTROUTING ACCEPT [125:25487]
:ndsALW - [0:0]
:ndsBLK - [0:0]
:ndsINC - [0:0]
:ndsOUT - [0:0]
:ndsTRU - [0:0]
-A PREROUTING -i wlan0 -j ndsOUT
-A PREROUTING -i wlan0 -j ndsBLK
-A PREROUTING -i wlan0 -j ndsTRU
-A POSTROUTING -o wlan0 -j ndsINC
-A ndsINC -d <REDACTED IP 1>/32 -j MARK --set-xmark 0x30000/0x30000
-A ndsINC -d <REDACTED IP 1>/32 -j ACCEPT
-A ndsOUT -s <REDACTED IP 1>/32 -m mac --mac-source <REDACTED MAC 1> -j MARK --set-xmark 0x30000/0x30000
-A ndsTRU -m mac --mac-source <REDACTED MAC 2> -j MARK --set-xmark 0x20000/0x20000
-A ndsTRU -m mac --mac-source <REDACTED MAC 3> -j MARK --set-xmark 0x20000/0x20000
COMMIT
# Generated by iptables-save v1.8.2 
*nat
:PREROUTING ACCEPT [319:25995]
:INPUT ACCEPT [105:8834]
:OUTPUT ACCEPT [76:5765]
:POSTROUTING ACCEPT [76:5765]
:ndsOUT - [0:0]
-A PREROUTING -i wlan0 -j ndsOUT
-A ndsOUT -d <REDACTED openNDS IP>/32 -p tcp -m tcp --dport 80 -j REDIRECT --to-ports 2050
-A ndsOUT -m mark --mark 0x20000/0x30000 -j RETURN
-A ndsOUT -m mark --mark 0x30000/0x30000 -j RETURN
-A ndsOUT -p udp -m udp --dport 67 -j ACCEPT
-A ndsOUT -p udp -m udp --dport 53 -j ACCEPT
-A ndsOUT -p tcp -m tcp --dport 53 -j ACCEPT
-A ndsOUT -p tcp -m tcp --dport 80 -j DNAT --to-destination <REDACTED openNDS IP>:2050
-A ndsOUT -j ACCEPT
COMMIT
```

```
cat /proc/sys/net/ipv4/ip_forward
1
```

On the client side I tried to 
```
~$ ping -c 10 8.8.8.8
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.

--- 8.8.8.8 ping statistics ---
10 packets transmitted, 0 recevied, 100% packet loss, time 9204ms

~$ timeout 10 host google.com || echo 'Timeout!'
Timeout!
```
and tried to access https://1.1.1.1 (which is an actual website that doesn't need dns to access), the browser fails after few seconds with **The connection has timed out**

Can someone help?

**Comment by bluewavenet** [comment]
@NourEddineX 
You seem to be well into a project but do not go into any detail of the basic configuration of your hardware.

**Can you verify you are complying with the basic prerequisites for running openNDS?**
ie see the documentation:
https://opennds.readthedocs.io/en/stable/install.html#prerequisites

Version 8.0.0 has been tested on x86-64 with OpenSuse Tumbleweed with no issues whatsoever.
It has also been tested on Raspbian Lite on Rpi as well as numerous platforms on OpenWrt, so I suspect you have a configuration issue rather than a bug....

It is not required to do anything at all with iptables as openNDS handles everything it needs, setting up BEFORE any other local firewall settings. Your ISP will not be able to see anything of openNDS as it will be behind the NAT of the router openNDS is running on.

**Please post details of your x86 router setup** including the private ip addresses you are using (there is no point in redacting these as they are, well, private and will be in use in millions of other places).

**Did it work with the basic configuration** before you made any changes?

**Comment by NourEddineX** [comment]
My router setup is ad follows:

eth0 (the wan iface) is manually configured ( 192.168.1.254/24 , gateway: 192.168.1.1 )
the link you provided mentions it should be configured via dhcp, I'm not sure if this mandatory, the gateway is the same as the provided by dhcp anyway, and the opennds router itself can access internet.

opennds is running on wlan0 (can be found in the configuration in the issue body)
wlan0 interface is 192.168.8.1/24 , dnsmasq configures dhcp in the range of 192.168.8.2-192.168.8.200

**Comment by NourEddineX** [comment]
also btw, it's an x86_64, i just wanted to state it isn't arm.

**Comment by NourEddineX** [comment]
I've tried to configure wan via dhcp, no luck as well
```
auto lo
iface lo inet loopback

allow-hotplug eth0
allow-hotplug wlan0

auto eth0
iface eth0 inet dhcp
# old conf below
#iface eth0 inet static
#       address 192.168.1.254/24
#       gateway 192.168.1.1

auto wlan0
iface wlan0 inet static
        hostapd /etc/hostapd/hostapd.conf
        address 192.168.8.1/24
        gateway 192.168.8.1
```

**Comment by bluewavenet** [comment]
@NourEddineX 

> configuration never worked in a way I could access internet

As clients are showing as authenticated, this implies there is an issue with the actual router config.

Can a client attached to wlan0 access the Internet if openNDS is **NOT** running?
ie run `systemctl stop opennds` then try connecting with a client.

Please also post the output of:
`ip route`

I assume your version of Debian is running something like firewalld. Do you have any firewall entries preventing a client on wlan0 from accessing the Internet?

**Comment by bluewavenet** [comment]
@NourEddineX 
DHCP is not mandatory on the wan interface - But:
It is a recommendation that makes everything much simpler.- it increases the probability that the interface is correctly configured in almost any situation.

**Comment by sicelo** [comment]
A quick look at your IP addressing seems suspect - your wan interface had its gateway as 192.168.1.1 according to your explanation and the now commented out /etc/network/interfaces block.

However, your wlan has this same address, and, not only that, but its gateway is also 192.168.1.1, that is, itself. The gateway for wlan0 should be the address for eth0, and you should have some NAT/Masquerading enabled and configured for the firewall

Additionally, the opennds config seems to indicate opennds is managing 0.0.0.0/0. You need to make that the same as your dnsmasq pool, i.e. 192.168.8.0/24

**Comment by bluewavenet** [comment]
@sicelo 
The wlan0 address seems to be 192.168.8.1 so should be ok but I think there is a misconfiguration here somewhere. 
`ip route` might give some indication...

`Managed IP range: 0.0.0.0/0`
This is the default and means "Manage the entire ip subnet of the Managed Interface" so is fine. 

> you should have some NAT/Masquerading enabled and configured for the firewall

You have "hit the nail on the head" :-D  I think this is most likely the problem! 
The ISP router will have no idea where 192.168.8.0/24 is, so clients will not receive anything.

I think if @NourEddineX  tries without openNDS running and still no client Internet access, then the lack of nat/masquerading will be more or less confirmed.

**Comment by NourEddineX** [comment]
```$ ip route
default via 192.168.1.1 dev eth0 
192.168.1.0/24 dev eth0 proto kernel scope link src 192.168.1.18 
192.168.8.0/24 dev wlan0 proto kernel scope link src 192.168.8.1 
```
this looks pretty normal

Also, I've added Masquerade, (sudo iptables -t nat -A POSTROUTING -o eth0 -j MASQUERADE) , this didn't same to help, even after stopping opennds, what do you think is missing?

**Comment by bluewavenet** [comment]
@NourEddineX 

> even after stopping opennds

So are you saying your router setup does not work with openNDS stopped?

> what do you think is missing?

I don't use Debian for anything so  I cannot say with any certainty, but maybe try running the following BEFORE starting openNDS:

```
echo 1 > /proc/sys/net/ipv4/ip_forward
/sbin/iptables -t nat -A POSTROUTING -o eth0 -j MASQUERADE
/sbin/iptables -A FORWARD -i eth0 -o wlan0 -m state --state RELATED,ESTABLISHED -j ACCEPT
/sbin/iptables -A FORWARD -i wlan0 -o eth0 -j ACCEPT
```

There must be a good tutorial somewhere on how to set up a NAT router on Debian......
Better still, why not put OpenWrt on your x86-64 :-D

**Comment by bluewavenet** [comment]
This one maybe:
https://gridscale.io/en/community/tutorials/debian-router-gateway/

**Comment by NourEddineX** [comment]
looks nice, it works, decpite I tried this very same process but didn't succeeded, I guess maybe the kernel needed a restarted
Also, I wasn't sure if opennds will handle the natting itself or I should configure manually, thank you anyway for helping with your precious time. :) have a good day/night

**Comment by bluewavenet** [comment]
@NourEddineX 
Excellent. It might be worthwhile for me to put this into the documentation.
Please feel free to open another issue if you have any questions.
